5] dice/dice.py

```diff
@@ -0,0 +1,101 @@
+"""
+Workshop_01 - Program_05
+----
+Dice
+----
+(c) Tomas Dolejsek 2024-01-30
+
+Board games and pen-and-paper RPG games use many types of dice, not just the well-known cubic ones.
+One of the most popular dice are, for example, a ten-sided one or even a hundred-sided one! Since dice are often tossed
+in games, writing "e.g. roll two ten-sided dice and add 20 to the result" each time would be boring, difficult,
+and wasting huge amounts of paper. In such situations, the code "roll 2D10+20" is used.
+
+The code for such a die has the following formula: xDy+z
+where:
+    * y – type of dice to use (e.g. D6, D10),
+    * x – number of dice rolled; for a single roll this parameter is omitted,
+    * z – modifier - number to add (or subtract) to the result of the roll (optional).
+
+Examples:
+
+2D10+10: roll 2 D10 dice, add 10 to the result,
+D6: a single roll of a cubic die,
+2D3: roll 2 three-sided dice,
+D12-1: roll one D12 die, subtract 1 from the result.
+Write a function that will:
+
+take such code as a string in the parameter,
+recognize all input data:
+    * type of dice,
+    * number of rolls,
+    * modifier,
+return an appropriate message if the given string is invalid,
+simulate the rolls and return the result.
+Types of dice used in games: D3, D4, D6, D8, D10, D12, D20, D100.
+"""
+
+import re
+from random import randint
+from collections import namedtuple
+DiceRoll = namedtuple('DiceRoll', ['number', 'dice', 'modifier'])
+
+
+def process_input(user: str) -> namedtuple or bool:
+    """
+    Take user's input and validate its correctness. If it's not correct, return False.
+    Using regex expression, extract needed information (number of rolls, type of dice and modifier).
+    Return the result as namedtuple.
+    :param user: string given by user
+    :return: False if incorrect input or namedtuple with extracted info (number of rolls, type of dice, modifier)
+    """
+    pattern = re.compile(r'^([0-9]*)D([0-9]+)(([+/-]?[0-9]+)?)$')
+    correct_dice = (3, 4, 6, 8, 10, 12, 20, 100)
+    matches = pattern.match(user)
+
+    if not matches:
+        return False
+    rolls = matches.group(1)  # everything before 'D' means number of rolls
+    dice = matches.group(2)  # type of dice
+    modifier = matches.group(3)  # the rest of the input is the modifier
+
+    rolls = int(rolls) if rolls else 1  # no number of rolls => rolls = 1
```

Comment:
tohle vsehcno jde jeste viz zjednodusit a zprahlednit. ale ano, groups jsou to reseni. jdes na to dobre